platform/core/lib/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+pub mod core {
+    use failure::CoreFailure;
+    use feature_categories;
+    use io::{file_system::DefaultFileSystem, key_value_store::DefaultKeyValueStore};
+    use preferences::Preferences;
+    use user_session::{TimeoutOption, UserSession, UserSessionTimeout};
+
+    use once_cell::sync::OnceCell;
+    use serde::{Deserialize, Serialize};
+    use std::sync::{Arc, Mutex};
+    use std::{sync::MutexGuard, time::Instant};
+
+    #[cfg(target_os = "android")]
+    use {
+        android_logger::Config,
+        log::{trace, Level},
+    };
+
+    // Core is held as a singleton.
+    static CORE: OnceCell<Mutex<Core>> = OnceCell::new();
+
+    #[derive(Debug, Clone, Serialize)]
+    pub enum PlatformRequest {
+        Example,
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    pub enum PlatformResponse {
+        Example(String),
+    }
+
+    pub trait PlatformHookRequest: Sync + Send {
+        fn perform_request(
+            &self,
+            request: PlatformRequest,
+        ) -> Result<PlatformResponse, CoreFailure>;
+    }
+
+    const PREFERENCES_DB: &str = "preferences_db";
+
+    // The Core would act as a composition root, containing any global configuration
+    // to be shared between components, as well managing components lifetime.
+    struct Core<'a> {
+        language_code: String,
+        #[allow(dead_code)]
+        preferences: Arc<Preferences>,
+        user_session: Mutex<UserSession<'a>>,
+        #[allow(dead_code)]
+        platform_hook: Box<dyn PlatformHookRequest>,
+    }
+
+    fn get_instance() -> Result<MutexGuard<'static, Core<'static>>, CoreFailure> {
```

Comment:
Is there any chance of the lifetime of `Core` being other than `'static`?

Reply:
No, Core is meant to live as long as the platform application does.

Reply:
So why is it parametrized by lifetime above? Couldn't you simply set lifetime as `static` in the definition of `Core`?

Reply:
No, it cannot, the lifetime specifier is simply special generic. One cannot use `'static` the same way one cannot have a generic struct with a concrete type specified.

I mean:
```rust
struct Core<'a> {
   let field: Usersession<'a>
}
```
Is conceptually the same as:
```rust
struct Core<T> {
   let field: UserSession<T>
}
```

Using `'static` at definition level would be equivalent to trying to have a generic like this:
```rust
struct Core<uint> {
   let field: UserSession<uint>
}
```

If you would try to use `'static` at definition level, I think Rust will throw the error that `'static` is a reserved lifetime specifier.